lib/fuser_mqubit.h

```diff
@@ -0,0 +1,1087 @@
+      next.qubits.reserve(8);
+      next.links.reserve(8);
+
+      GetAddedQubits(pgate1, pgate2, *ngate, next);
+
+      if (cur_size + next.qubits.size() > max_fused_size) continue;
+
+      scratch.push_back(std::move(next));
+      next_gates.push_back(&scratch.back());
+    }
+  }
+
+  static bool GetPrevAvailableGates(unsigned max_fused_size,
+                                    unsigned cur_size, unsigned level,
+                                    const GateF& ngate1, const GateF* ngate2,
+                                    std::vector<GateA>& scratch,
+                                    std::vector<GateA*>& prev_gates) {
+    prev_gates.resize(0);
+
+    for (auto link : ngate1.links) {
+      if (link->prev == nullptr) continue;
+
+      auto pgate = link->prev->val;
+
+      if (pgate->visited > 997 || pgate->visited == level) continue;
+
+      if (pgate->visited > 1 || pgate->parent->unfusible) {
+        prev_gates.resize(0);
+        return false;
+      }
+
+      GateA prev = {pgate, {}, {}};
+      prev.qubits.reserve(8);
+      prev.links.reserve(8);
+
+      GetAddedQubits(ngate1, ngate2, *pgate, prev);
+
+      bool all_prev_visited = true;
+
+      for (auto link : pgate->links) {
+        if (link->prev == nullptr) continue;
+
+        if (link->prev->val->visited <= 997) {
+          all_prev_visited = false;
+          break;
+        }
+      }
+
+      if (!all_prev_visited) {
+        prev_gates.resize(0);
+        return false;
+      }
+
+      if (cur_size + prev.qubits.size() > max_fused_size) continue;
+
+      if (all_prev_visited) {
+        scratch.push_back(std::move(prev));
+        prev_gates.push_back(&scratch.back());
+      }
+    }
+
+    return true;
+  }
+
+  static void GetAddedQubits(const GateF& fgate0, const GateF* fgate1,
+                             const GateF& fgate2, GateA& added) {
+    for (std::size_t i = 0; i < fgate2.qubits.size(); ++i) {
+      unsigned q2 = fgate2.qubits[i];
+
+      bool have_added_qubits = true;
+
+      for (auto q0 : fgate0.qubits) {
+        if (q0 == q2) {
+          have_added_qubits = false;
+          break;
+        }
+      }
+
+      if (have_added_qubits && fgate1 != nullptr) {
+        for (auto q1 : fgate1->qubits) {
+          if (q1 == q2) {
+            have_added_qubits = false;
+            break;
+          }
+        }
+      }
+
+      if (have_added_qubits) {
+        added.qubits.push_back(q2);
+        added.links.push_back(fgate2.links[i]);
+      }
+    }
+  }
+
+  // Fuse smaller gates with fgate back in gate time.
+  static void FusePrev(unsigned level, GateF& fgate) {
+    std::vector<const Gate*> gates;
+    gates.reserve(fgate.gates.capacity());
+
+    uint64_t bad_mask = 0;
+    auto links = fgate.links;
+
+    bool may_have_gates_to_fuse = true;
+
+    while (may_have_gates_to_fuse) {
+      may_have_gates_to_fuse = false;
+
+      std::sort(links.begin(), links.end(),
+                [](const Link* l, const Link* r) -> bool {
```

Comment:
This lambda is reused in FuseNext - it can probably be separated out as a helper method.

Alternatively, it seems like FusePrev and FuseNext are very similar - the shared portion may be possible to extract into a helper method.

Reply:
Added a helper method.